### ISSUE TITLE
Java generics support

### ISSUE THREAD

**Issue by beikov** [smallrye/smallrye-graphql#574]
I was hoping that `io.smallrye.graphql.schema.model.Reference#parametrizedTypeArguments` would already capture the type variable assignments, but it doesn't. Why not? Generics support is really crucial for libraries that try to add support for standard containers like the relay `Connection`, `Edge` etc.

Currently, a method return type like `Connection<MyType>` is interpreted as a `Field` with `Wrapper(Connection)` and `Reference(String)`.

**Comment by phillip-kruger** [comment]
@velias can you comment ? Thanks

**Comment by velias** [comment]
@beikov I'm not sure if I fully understand your question. Can you please specify at which point are you looking at the model? And maybe provide some example code. `Wrapper` has been added relatively recently (after support for generics), we talked with @phillip-kruger that generics have to be aligned with Wrappers and better model decoupling from the scanning process, but it is not done yet as I understand it. So your requirements can be important to do it correctly.

**Comment by beikov** [comment]
Hi @velias,

I simply have a method like this:

```
@GraphqlApi
class Persons {
  @Query("persons")
  public RelayConnection<Person> persons() {
    //...
  }
}
```

and was getting weird errors on startup. I inspected the model by listening to the fired `Operation` CDI events and inspected the structure. Not sure how much the RelayConnection type plays in here, but fwiw here is the definition:

```
class GraphQLRelayConnection<T> implements Serializable {

    public List<GraphQLRelayEdge<T>> getElements() {
        // ...
    }

    public RelayPageInfo getPageInfo() {
        // ...
    }

    public long getTotalCount() {
        // ...
    }
```

**Comment by velias** [comment]
Thanks, and which version of the smallrye-graphql do you run and how? In Quarkus, which version of quarkus in this case please? 
I'm not sure at which point is `Operation` CDI event fired, so I have to investigate more, but want to do it in the exact version as you

**Comment by beikov** [comment]
I tried with Quarkus 1.10.3.Final. The event is fired during bootstrap in `io.smallrye.graphql.bootstrap.Bootstrap#generateGraphQLSchema`.

**Comment by velias** [comment]
Hi @beikov, I checked `Operation` events in my example app and everything looks good to me. So I'm not sure where your problem comes from. Do you face the problem in `dev` mode or another (eg. native) run mode? Can you provide me reproducer app please (maybe the problem relates to the structure of your model classes/interfaces). 
You also mentioned "weird errors on startup", any details for them? Thanks in advance

**Comment by beikov** [comment]
I played around in https://github.com/beikov/blaze-persistence/tree/issues/1125 the test project/module is https://github.com/beikov/blaze-persistence/tree/issues/1125/examples/microprofile-graphql

Startup errors were something about can not implement interface or something like that. I only tried the JVM mode so far.

**Comment by velias** [comment]
After some playing around I just realized this error message in log when quarkus app is started:
```
2020-12-18 12:22:41,428 WARN  [io.sma.gra.sch.cre.ReferenceCreator] (build-26) Class [com.blazebit.persistence.integration.graphql.GraphQLRelayConnection] in not indexed in Jandex. Can not scan Object Type, defaulting to String Scalar
```
It explains why the reference is not resolved. `GraphQLRelayConnection` class comes from jar which is in maven dependency, it is seen by Quarkus during startup, but not seen during Schema build (which is done by Maven if I'm sure). I'm not fully sure how Jandex index is populated for the build and how to patch this problem. @phillip-kruger any idea, have you ever tested schema build with classes coming from dependencies?

**Comment by phillip-kruger** [comment]
No, the handed index is provided, not sure if we can tell Quarkus to included that somehow ?

**Comment by velias** [comment]
@phillip-kruger BTW we should maybe change this problem from WARN to ERROR and terminate schema build with error, as the message can be overseen easily and runtime schema is strange then and not os easy to detect where the problem is?

**Comment by beikov** [comment]
Can I somehow create this Jandex index through a Maven plugin? I wouldn't mind as long as it makes it work :)

**Comment by velias** [comment]
@beikov  according to https://stackoverflow.com/questions/55513502/how-to-create-a-jandex-index-in-quarkus-for-classes-in-a-external-module

Add this to application.properties:
```
quarkus.index-dependency.blaze-persistence-integration-graphql.group-id=com.blazebit
quarkus.index-dependency.blaze-persistence-integration-graphql.artifact-id=blaze-persistence-integration-graphql
```
Reference is resolved correctly then

**Comment by beikov** [comment]
Thanks, I'll try that out!

**Comment by velias** [comment]
@phillip-kruger even after this configuration I see 
```
2020-12-18 12:48:29,491 WARN  [io.sma.gra.sch.cre.ReferenceCreator] (build-8) Class [java.io.Serializable] in not indexed in Jandex. Can not scan Object Type, defaulting to String Scalar
```
in the log as one of model interfaces extends this interface. I think model parser should ignore this interface as there is not any point to have it in GraphQL schema. WDYT?

**Comment by beikov** [comment]
Sounds reasonable. I think any interface without abstract methods (neither declared nor inherited) should be ignored. Not sure what the point would be to introduce an interface in the graphql schema for such a Java construct.

**Comment by velias** [comment]
@beikov aha, `Serializable` interface is evaluated because `com.blazebit.persistence.integration.graphql.GraphQLCursor.getTuple()` returns array of `Serializable`. So either change it to something else, or try somehow ask Quarkus to add this interface into Jandex index

**Comment by beikov** [comment]
I added the jandex plugin and also a beans.xml. The generics handling seems to work now, but I hit another roadblock. Apparently, the GraphQL interface types (that implement itself?) are added for Java interfaces. I'm seeing the following errors now:

```
interface type 'CatWithOwnerView' cannot implement 'CatWithOwnerView' because that would result on a circular reference
interface type 'PersonSimpleView' cannot implement 'PersonSimpleView' because that would result on a circular reference
```

I updated the branch I posted before so you can take a look: https://github.com/beikov/blaze-persistence/tree/issues/1125/examples/microprofile-graphql

Since this type is an "output type" i.e. it is only used as return type for a query, I don't see a reason why GraphQL needs an implementation. I actually tried to provide an implementation as you can see in the code I commented out in `com.blazebit.persistence.examples.microprofile.graphql.GraphQLProducer#configure` but it's not possible to do this in a generic way as I have no access to the fields the interface type defines. Is this maybe a bug?

**Comment by beikov** [comment]
Shouldn't Java interfaces only be modeled as GraphQL interface if I annotate it with `@Interface` as described in the MP-GraphQL specification? https://download.eclipse.org/microprofile/microprofile-graphql-1.0/microprofile-graphql.html#_graphql_interfaces

I also tried annotating the interfaces with `@Type` but that didn't help either.

**Comment by velias** [comment]
@beikov I tried in my project, interface without implementation for output is OK, it doesn't fail.

I think your error says something a bit different. BTW it comes from schema validation implemented in `graphql.schema.validation`which is 'graphgl-java' library, it doesn't come from smallrye-graphql itself.
Anyway, the error says something about circular dependency, looks like in schema obtained from introspection interface `CatWithOwnerView` implements/extends itself. I'm not sure where it comes from, I do not see anything strange in the model java object itself. Does your framework perform any manipulation with the graphql schema during its creation through introspection?

**Comment by beikov** [comment]
Hey @velias,

no there are no transformations yet. I'm just trying to get a minimal example running. The type is represented as interface in the quarkus built smallrye schema already which AFAIK runs as build item, right? Maybe the issue is there?

**Comment by velias** [comment]
OK, looks like I found error in `io.smallrye.graphql.schema.creator.type.InterfaceCreator.addInterfaces(InterfaceType, ClassInfo)`. Looks like resolution of the interface inheritance is broken, and interface itself is added as parent interface instead of real parent interface. I'll try to replicate it in my project and in the smallrye-grpahql tests itself, if yes then I'll open another bug to track it separately and patch it in master.

**Comment by beikov** [comment]
Thanks so much @velias!

I hope this can be fixed soon and update the version in Quarkus so that I can continue my experiments :)

**Comment by velias** [comment]
@beikov closing this issue as the original topic has been resolved and we have new issue #579 for later problem.
